examples/e14_global_data/src/main.rs

```diff
@@ -0,0 +1,215 @@
+//! In this example, you will be shown various ways of sharing data between events and commands.
+//! And how to use locks correctly to avoid deadlocking the bot.
+
+use std::{sync::{Arc, atomic::{AtomicUsize, Ordering}}, collections::HashMap, env};
+
+use serenity::{
+    async_trait,
+    model::{channel::Message, gateway::Ready},
+    prelude::*,
+
+    framework::standard::{
+        Args, CommandResult, StandardFramework,
+        macros::{command, group, hook},
+    },
+};
+
+use tokio::sync::RwLock;
+
+// A container type is created for inserting into the Client's `data`, which
+// allows for data to be accessible across all events and framework commands, or
+// anywhere else that has a copy of the `data` Arc.
+// These places are usually where either Context or Client is present.
+//
+// Documentation about TypeMap can be found here:
+// https://docs.rs/serenity/0.9.0-rc.2/serenity/prelude/struct.TypeMap.html
+struct CommandCounter;
+
+impl TypeMapKey for CommandCounter {
+    type Value = Arc<RwLock<HashMap<String, u64>>>;
+}
+
+struct MessageCount;
+
+impl TypeMapKey for MessageCount {
+    // While you will be using RwLock or Mutex most of the times you want to modify data,
+    // sometimes it's not required; Like for example, with static data, or if you are using other
+    // kinds sof atomic operators.
+    //
+    // Arc shoud stay, to allow for the data lock to be closed early.
+    type Value = Arc<AtomicUsize>;
+}
+
+#[group]
+#[commands(ping, command_usage, owo_count)]
+struct General;
+
+#[hook]
+async fn before(ctx: &Context, msg: &Message, command_name: &str) -> bool {
+    println!("Running command '{}' invoked by '{}'", command_name, msg.author.tag());
+
+    let counter_lock = {
+        // While data is a RwLock, it's recommended that you always open the lock as read.
+        // this is mainly done to avoid Deadlocks for having a possible writer waiting for multiple
+        // readers to close.
+        let data_read = ctx.data.read().await;
+
+        // Since the CommandCounter Value is wrapped in an Arc, cloning will not duplicate the
+        // data, instead the reference is cloned.
+        // We wap every value on an arc, as to keep the data lock open for the least time possible,
+        // to again, avoid Deadlocking it.
+        data_read.get::<CommandCounter>().expect("Expected CommandCounter in TypeMap.").clone()
+    };
+
+    // Just like with client.data in main, we want to keep write locks open the least time
+    // possible, so we wrap them on a block so they get automatically closed at the end.
+    {
+        // The HashMap of CommandCounter is wrapped on RwLock; Since we want to write to it, we will
```

Comment:
Ditto. And additionally, the hashmap is contained *in* an `RwLock`, not placed on its top or somewhere on its head.
```suggestion
        // The HashMap of CommandCounter is wrapped in an RwLock; since we want to write to it, we will
```